File: uidesigner/com.nokia.carbide.cpp.uiq.ui/src/com/nokia/carbide/cpp/uiq/ui/vieweditor/CommandId.java
/*
* Copyright (c) 2009 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of the License "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description: 
*
*/
/* START_USECASES: CU6 END_USECASES */
package com.nokia.carbide.cpp.uiq.ui.vieweditor;
/**
 * This class represents the Command Id in the local model.
 *
 */
public class CommandId {
	private String name;
	private String commandId;
	private String commandEvent;
	private int references = 0;
	
	/**
	 * Class constructor
	 * @param name
	 */
	public CommandId(String name){
		this.setName(name);		
	}
	/**
	 * Sets the name for the command id
	 * @param name
	 */
	public void setName(String name) {
		this.name = name;		
	}
	/**
	 * Returns the name of the command id
	 * @return
	 */
	public String getName() {
		return name;
	}
	/**
	 * Set the property commandid
	 * @param commandId
	 */
	public void setCommandId(String commandId) {
		this.commandId = commandId;
	}
	/**
	 * Returns the property command id
	 * @return
	 */
	public String getCommandId() {
		return commandId;
	}
	/**
	 * Sets the command Event property
	 * @param commandEvent
	 */
	public void setCommandEvent(String commandEvent) {
		this.commandEvent = commandEvent;
	}
	/**
	 * Returns the command Event property
	 * @return
	 */
	public String getCommandEvent() {
		return commandEvent;
	}
	
	/**
	 * Add a reference for the commandID
	 * @param adding
	 */
	public void refreshReference(int adding){
		references = references + adding;
	}
	/**
	 * Returns the number of references.
	 * @return number of references
	 */
	public int getReference(){
		return references;
	}

}

File: uidesigner/com.nokia.carbide.cpp.uiq.components/components/command_related/CommandId/CommandId.js
/*
* Copyright (c) 2007 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of the License "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description: 
* START_USECASES: CU11 END_USECASES
*
*
*/

include("../../containerLibrary.js")
include("../../renderLibrary.js")


function CommandId() {
}


////////////////////////////////////////////////////////////////////////////////
//IComponentValidator

CommandId.prototype.validate = function(instance) {
	if (instance != null || instance != "") {
		if(instance.properties.commandId != "" && instance.properties.systemCommandId != "" ){
			var modelMessage = newModelMessage(IStatus.WARNING,formatString(lookupString("validate.flags"),
					[instance.name ]),instance, "flags", null);
			return [ modelMessage ];												
		}
	}
	return null;
}

CommandId.prototype.queryPropertyChange = function(instance, propertyId, newValue) {
	return null;
}
///////////////////////////////////////////////////////////////////////////////
//IInitializer
CommandId.prototype.initialize = function(instance, isConfigured) {
	if (isConfigured) 
		return;
	var initialCommandIdProposal = instance.properties.name;
	var commandListInstance = instance.parent;
	var tryCounter = 0;
	var finalCommandId = getValidCommandId(instance.properties.name, commandListInstance, initialCommandIdProposal, tryCounter);
	
	/*if (isConfigured && (finalCommandId != initialCommandIdProposal || instance.properties.systemCommandId.length > 0)) {
		return;
	}*/
	
	instance.properties.commandId = finalCommandId;
	instance.properties.handlerNameTemplate = instance.properties.commandId;
}

function getValidCommandId(commandIdName, commandList, initialCommandIdProposal, tryCounter) {
	/*
	println("------------" + tryCounter + "------------");
	println("commandIdName=" + commandIdName);
	println("initialCommandIdProposal=" + initialCommandIdProposal);
	println("tryCounter=" + tryCounter);
	println("initialCommandIdProposal + tryCounter=" + initialCommandIdProposal + tryCounter);
	*/
	
	var RECURSIVE_LIMIT = 100;
	if (tryCounter > RECURSIVE_LIMIT)
		return initialProposal;
		
	var commandIdProposal;	
	if (tryCounter != 0) {
		commandIdProposal = initialCommandIdProposal + tryCounter;
	} else {
		commandIdProposal = initialCommandIdProposal;
	}
	
	var repeated = false;
	for (var i=0; i<commandList.children.length; i++) {
		var childProperties = commandList.children[i].properties; 
		if(childProperties.name != commandIdName && childProperties.commandId == commandIdProposal) {
			repeated = true;
			break;
		}
	}
	
	if (repeated) {
		return getValidCommandId(commandIdName, commandList, initialCommandIdProposal, ++tryCounter);
	} else {
		return commandIdProposal;
	}
}


////////////////////////////////////////////////////////////////////////////////
//IPropertyListener

CommandId.prototype.propertyChanged = function(instance, property) {
	if (  (property=="commandId" && instance.properties.systemCommandId == "" && instance.properties.commandId != "")
		||(property=="systemCommandId" && instance.properties.systemCommandId == "" && instance.properties.commandId != "") ){
		instance.properties.handlerNameTemplate=instance.properties.commandId;
	}
	if (  (property=="commandId" && instance.properties.systemCommandId != "" && instance.properties.commandId == "")
			||(property=="systemCommandId" && instance.properties.systemCommandId != "" && instance.properties.commandId == "") ){
			instance.properties.handlerNameTemplate=instance.properties.systemCommandId;
	}
	if (  (property=="commandId" && instance.properties.systemCommandId != "" && instance.properties.commandId != "")
			||(property=="systemCommandId" && instance.properties.systemCommandId != "" && instance.properties.commandId != "") ){
			instance.properties.handlerNameTemplate=instance.properties.commandId;
	}

}
